ci/test.sh

```diff
@@ -13,6 +13,9 @@ chown -R ubuntu:ubuntu /cache/
 echo '127.0.1.11 internal-stargate.local' >> /etc/hosts
 echo '127.0.1.12 internal-stargate.local' >> /etc/hosts
 
+# restart docker
+service docker --full-restart
```

Comment:
It's not clear to me why this fixes the issue. Maybe the comment should say something about why this is done?

Reply:
I am also curious. It seems like this would only run once (for each backend test), and not between IT classes.
If so, its effect would be quite limited. Tests themselves use, I think (based on log output) testcontainers to start docker containers for individual tests.

Reply:
I believe this is a good approach that we need to attempt. We do not know that it will work but have reason to believe it should help based on the output we observed from build logs. Here is an example of a portion of the output from executing one of the JWT test cases that is executing Keycloak in Docker using testcontainers:

```
Step #4 - "cassandra-3.11": [INFO] Running io.stargate.it.cql.JwtAuthTest
Step #4 - "cassandra-3.11": INFO  [main] 2022-01-19 16:38:42,500 ExternalStorage.java:100 - Reusing matching storage cluster @io.stargate.it.storage.ClusterSpec(shared=true, nodes=1) for [engine:junit-jupiter]/[class:io.stargate.it.cql.JwtAuthTest]
Step #4 - "cassandra-3.11": WARN  [main] 2022-01-19 16:38:42,573 TestcontainersConfiguration.java:309 - Attempted to read Testcontainers configuration file at file:/home/ubuntu/.testcontainers.properties but the file was not found. Exception message: FileNotFoundException: /home/ubuntu/.testcontainers.properties (No such file or directory)
Step #4 - "cassandra-3.11": INFO  [main] 2022-01-19 16:38:42,583 DockerMachineClientProviderStrategy.java:58 - docker-machine executable was not found on PATH ([/usr/local/sbin, /usr/local/bin, /usr/sbin, /usr/bin, /sbin, /bin, /usr/lib/jvm/java-1.8.0-openjdk-amd64/bin])
Step #4 - "cassandra-3.11": INFO  [main] 2022-01-19 16:38:43,061 DockerClientProviderStrategy.java:158 - Found Docker environment with local Unix socket (unix:///var/run/docker.sock)
Step #4 - "cassandra-3.11": INFO  [main] 2022-01-19 16:38:43,081 DockerClientFactory.java:180 - Docker host IP address is 172.17.0.1
Step #4 - "cassandra-3.11": INFO  [main] 2022-01-19 16:38:43,109 DockerClientFactory.java:192 - Connected to docker: 
Step #4 - "cassandra-3.11":   Server Version: 20.10.12
Step #4 - "cassandra-3.11":   API Version: 1.41
Step #4 - "cassandra-3.11":   Operating System: Ubuntu 18.04.6 LTS
Step #4 - "cassandra-3.11":   Total Memory: 28881 MB
Step #4 - "cassandra-3.11": INFO  [main] 2022-01-19 16:38:43,109 DockerClientFactory.java:221 - Checking the system...
Step #4 - "cassandra-3.11": INFO  [main] 2022-01-19 16:38:43,110 DockerClientFactory.java:295 - ?? Docker server version should be at least 1.6.0
Step #4 - "cassandra-3.11": INFO  [main] 2022-01-19 16:38:43,113 ImageNameSubstitutor.java:50 - Image name substitution will be performed by: DefaultImageNameSubstitutor (composite of 'ConfigurationFileImageNameSubstitutor' and 'PrefixingImageNameSubstitutor')
Step #4 - "cassandra-3.11": INFO  [main] 2022-01-19 16:38:43,138 RegistryAuthLocator.java:148 - Failure when attempting to lookup auth config. Please ignore if you don't have images in an authenticated registry. Details: (dockerImageName: alpine:3.5, configFile: /home/ubuntu/.docker/config.json. Falling back to docker-java default behaviour. Exception message: /home/ubuntu/.docker/config.json (No such file or directory)
Step #4 - "cassandra-3.11": INFO  [docker-java-stream-1239536715] 2022-01-19 16:38:43,410 LoggedPullImageResultCallback.java:38 - Starting to pull image
```

And here is what it looks like on an unsuccessful execution:

```
Step #4 - "cassandra-4.0": [INFO] Running io.stargate.it.cql.JwtAuthTest
Step #4 - "cassandra-4.0": INFO  [main] 2022-01-19 16:41:24,566 ExternalStorage.java:100 - Reusing matching storage cluster @io.stargate.it.storage.ClusterSpec(shared=true, nodes=1) for [engine:junit-jupiter]/[class:io.stargate.it.cql.JwtAuthTest]
Step #4 - "cassandra-4.0": WARN  [main] 2022-01-19 16:41:24,648 TestcontainersConfiguration.java:309 - Attempted to read Testcontainers configuration file at file:/home/ubuntu/.testcontainers.properties but the file was not found. Exception message: FileNotFoundException: /home/ubuntu/.testcontainers.properties (No such file or directory)
Step #4 - "cassandra-4.0": INFO  [main] 2022-01-19 16:41:24,660 DockerMachineClientProviderStrategy.java:58 - docker-machine executable was not found on PATH ([/usr/local/sbin, /usr/local/bin, /usr/sbin, /usr/bin, /sbin, /bin, /usr/lib/jvm/java-1.8.0-openjdk-amd64/bin])
Step #4 - "cassandra-4.0": WARN  [s13-admin-0] 2022-01-19 16:41:28,245 Loggers.java:39 - [s13|/127.0.2.8:9043]  Error while opening new channel (ConnectionInitException: [s13|connecting...] Protocol initialization request, step 1 (STARTUP {CQL_VERSION=3.0.0, DRIVER_NAME=DataStax Java driver for Apache Cassandra(R), DRIVER_VERSION=4.9.0, CLIENT_ID=2c3800eb-2348-4c1e-91a1-5d4d6dd9c026}): failed to send request (java.nio.channels.ClosedChannelException))
Step #4 - "cassandra-4.0": WARN  [s14-admin-1] 2022-01-19 16:41:29,905 Loggers.java:39 - [s14|/127.0.2.7:29042]  Error while opening new channel (ConnectionInitException: [s14|connecting...] Protocol initialization request, step 1 (STARTUP {CQL_VERSION=3.0.0, DRIVER_NAME=DataStax Java driver for Apache Cassandra(R), DRIVER_VERSION=4.9.0, CLIENT_ID=27ba8be7-10ce-4667-b286-dcbdeaa2b946}): failed to send request (java.nio.channels.ClosedChannelException))
Step #4 - "cassandra-4.0": WARN  [s15-admin-0] 2022-01-19 16:41:30,203 Loggers.java:39 - [s15|/127.0.2.8:9043]  Error while opening new channel (ConnectionInitException: [s15|connecting...] Protocol initialization request, step 1 (STARTUP {CQL_VERSION=3.0.0, DRIVER_NAME=DataStax Java driver for Apache Cassandra(R), DRIVER_VERSION=4.9.0, CLIENT_ID=89705602-4a25-4c16-a943-10a2e20b859a}): failed to send request (java.nio.channels.ClosedChannelException))
Step #4 - "cassandra-4.0": WARN  [s14-admin-0] 2022-01-19 16:41:30,234 Loggers.java:39 - [s14|/127.0.2.8:9043]  Error while opening new channel (ConnectionInitException: [s14|connecting...] Protocol initialization request, step 1 (STARTUP {CQL_VERSION=3.0.0, DRIVER_NAME=DataStax Java driver for Apache Cassandra(R), DRIVER_VERSION=4.9.0, CLIENT_ID=27ba8be7-10ce-4667-b286-dcbdeaa2b946}): failed to send request (java.nio.channels.ClosedChannelException))
Step #4 - "cassandra-4.0": WARN  [s13-admin-1] 2022-01-19 16:41:34,207 Loggers.java:39 - [s13|/127.0.2.7:9043]  Error while opening new channel (ConnectionInitException: [s13|connecting...] Protocol initialization request, step 1 (STARTUP {CQL_VERSION=3.0.0, DRIVER_NAME=DataStax Java driver for Apache Cassandra(R), DRIVER_VERSION=4.9.0, CLIENT_ID=2c3800eb-2348-4c1e-91a1-5d4d6dd9c026}): failed to send request (java.nio.channels.ClosedChannelException))
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2022-01-19 16:41:35,316 ProcessRunner.java:78 - storage0-0> INFO  [Messaging-EventLoop-3-16] 2022-01-19 16:41:35,315 NoSpamLogger.java:92 - /127.0.0.1:7000->/127.0.2.3:7000-URGENT_MESSAGES-[no-channel] failed to connect
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2022-01-19 16:41:35,316 ProcessRunner.java:78 - storage0-0> io.netty.channel.AbstractChannel$AnnotatedConnectException: finishConnect(..) failed: Connection refused: /127.0.2.3:7000
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2022-01-19 16:41:35,316 ProcessRunner.java:78 - storage0-0> Caused by: java.net.ConnectException: finishConnect(..) failed: Connection refused
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2022-01-19 16:41:35,316 ProcessRunner.java:78 - storage0-0> 	at io.netty.channel.unix.Errors.throwConnectException(Errors.java:124)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2022-01-19 16:41:35,316 ProcessRunner.java:78 - storage0-0> 	at io.netty.channel.unix.Socket.finishConnect(Socket.java:251)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2022-01-19 16:41:35,316 ProcessRunner.java:78 - storage0-0> 	at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.doFinishConnect(AbstractEpollChannel.java:673)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2022-01-19 16:41:35,316 ProcessRunner.java:78 - storage0-0> 	at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.finishConnect(AbstractEpollChannel.java:650)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2022-01-19 16:41:35,316 ProcessRunner.java:78 - storage0-0> 	at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.epollOutReady(AbstractEpollChannel.java:530)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2022-01-19 16:41:35,316 ProcessRunner.java:78 - storage0-0> 	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:470)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2022-01-19 16:41:35,316 ProcessRunner.java:78 - storage0-0> 	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:378)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2022-01-19 16:41:35,316 ProcessRunner.java:78 - storage0-0> 	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2022-01-19 16:41:35,316 ProcessRunner.java:78 - storage0-0> 	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2022-01-19 16:41:35,316 ProcessRunner.java:78 - storage0-0> 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2022-01-19 16:41:35,316 ProcessRunner.java:78 - storage0-0> 	at java.lang.Thread.run(Thread.java:748)
Step #4 - "cassandra-4.0": WARN  [s5-admin-1] 2022-01-19 16:41:50,018 Loggers.java:39 - [s5|/127.0.2.2:9043]  Error while opening new channel (ConnectionInitException: [s5|connecting...] Protocol initialization request, step 1 (STARTUP {CQL_VERSION=3.0.0, DRIVER_NAME=DataStax Java driver for Apache Cassandra(R), DRIVER_VERSION=4.9.0, CLIENT_ID=2b4072dc-eba8-4171-86d2-fd4f47dc7cf8}): failed to send request (java.nio.channels.ClosedChannelException))
Step #4 - "cassandra-4.0": ERROR [main] 2022-01-19 16:41:55,001 DockerClientProviderStrategy.java:208 - Could not find a valid Docker environment. Please check configuration. Attempted configurations were:
Step #4 - "cassandra-4.0": ERROR [main] 2022-01-19 16:41:55,002 DockerClientProviderStrategy.java:210 -     UnixSocketClientProviderStrategy: failed with exception TimeoutException (Timeout waiting for result with exception). Root cause IOException (native connect() failed : Permission denied)
Step #4 - "cassandra-4.0": ERROR [main] 2022-01-19 16:41:55,002 DockerClientProviderStrategy.java:212 - As no valid configuration was found, execution cannot continue
Step #4 - "cassandra-4.0": [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 30.433 s <<< FAILURE! - in io.stargate.it.cql.JwtAuthTest
```

From the comparison we deduce that the Docker daemon is not running on the unsuccessful case.

Reply:
Ok I am just trying to understand how -- considering that this is a transient issue, test could pass on CI with no changes at all. I am not saying it could not be a proper fix, just trying to know the logic.
I agree that the problem seems to relate to something in docker server state being bad (like server not even running). But why does it only seem to fail these particular tests, and not others?

Reply:
The tests that fail are the ones that rely on testcontainers to launch Keycloak in Docker.

Reply:
Ahhhh. So no other tests use testcontainers? And hence nothing else would be affected.

Reply:
I am sorry for the confusion and lack of the context.. Will try to be more descriptive.. I always do that, but here I was testing different things and forgot..

My analysis:
* I started with the `Could not find a valid Docker environment` error message, googled that together with `testcontainers` and got to this thread https://stackoverflow.com/questions/61108655/test-container-test-cases-are-failing-due-to-could-not-find-a-valid-docker-envi
* I focused around this comment `In my case this error caused by root permission. If you are using ubuntu, docker wants permission to create container and testcontainers can not ask you for this permission at runtime. So you need to make docker runnable without sudo.` and checked the `Dockerfile` on how we install Docker and this seemed fine to me: 
  ```
  RUN adduser --disabled-password --gecos "" ubuntu \
      && usermod -aG sudo ubuntu \
      && echo "ubuntu     ALL=(ALL) NOPASSWD:/usr/bin/python" >> /etc/sudoers \
      && usermod -aG docker ubuntu \
      && su - ubuntu
  ```
* Re-checked my own clean-Ubuntu scripts for installing everything and noticed that I had that docker restart after I added a user to a docker group (which was missing here)
* I wanted to see what happens if the docker is restarted before test run.. Seems docker was not started at all.. My thinking was that this could be the issue..

With that:
* Is this solving the issue? No idea.
* How it was working for some CIs and for others not? No idea.
* Will the problem occur again? No idea.

I can fire the build again, just to try to hit that failure case.

`/gcbrun`

Reply:
/gcbrun

Reply:
Why the hack is new build not started.. Will re-run manually..

Reply:
@ivansenic Thank you very much for the detail explanation! This sounds like a reasonable thing to try then.